index.js

```diff
@@ -1,2 +1,601 @@
+
+  // digest the options
+  var digest = options.digest || exports.digest(options);
+
+  // compute HOTP offset
+  var offset = digest[digest.length - 1] & 0xf;
+
+  // calculate binary code (RFC4226 5.4)
+  var code = (digest[offset] & 0x7f) << 24 |
+    (digest[offset + 1] & 0xff) << 16 |
+    (digest[offset + 2] & 0xff) << 8 |
+    (digest[offset + 3] & 0xff);
+
+  // left-pad code
+  code = new Array(digits + 1).join('0') + code.toString(10);
+
+  // return length number off digits
+  return code.substr(-digits);
+};
+
+// Alias counter() for hotp()
+exports.counter = exports.hotp;
+
+/**
+ * Verify a counter-based one-time token against the secret and return the delta.
+ * By default, it verifies the token at the given counter value, with no leeway
+ * (no look-ahead or look-behind). A token validated at the current counter value
+ * will have a delta of 0.
+ *
+ * You can specify a window to add more leeway to the verification process.
+ * Setting the window param will check for the token at the given counter value
+ * as well as `window` tokens ahead (one-sided window). See param for more info.
+ *
+ * `verifyDelta()` will return the delta between the counter value of the token
+ * and the given counter value. For example, if given a counter 5 and a window
+ * 10, `verifyDelta()` will look at tokens from 5 to 15, inclusive. If it finds
+ * it at counter position 7, it will return `{ delta: 2 }`.
+ *
+ * @param {Object} options
+ * @param {String} options.secret Shared secret key
+ * @param {String} options.token Passcode to validate
+ * @param {Integer} options.counter Counter value. This should be stored by
+ *   the application and must be incremented for each request.
+ * @param {Integer} [options.digits=6] The number of digits for the one-time
+ *   passcode.
+ * @param {Integer} [options.window=0] The allowable margin for the counter.
+ *   The function will check "W" codes in the future against the provided
+ *   passcode, e.g. if W = 10, and C = 5, this function will check the
+ *   passcode against all One Time Passcodes between 5 and 15, inclusive.
+ * @param {String} [options.encoding="ascii"] Key encoding (ascii, hex,
+ *   base32, base64).
+ * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
+ *   sha512).
+ * @return {Object} On success, returns an object with the counter
+ *   difference between the client and the server as the `delta` property (i.e.
+ *   `{ delta: 0 }`).
+ * @method hotp․verifyDelta
+ * @global
+ */
+
+exports.hotp.verifyDelta = function hotpVerifyDelta (options) {
+  var i;
+
+  // shadow options
+  options = Object.create(options);
+
+  // unpack options
+  var token = options.token;
+  var window = parseInt(options.window || 0, 10);
+  var counter = parseInt(options.counter || 0, 10);
+
+  // loop from C to C + W
+  for (i = counter; i <= counter + window; ++i) {
+    options.counter = i;
+    if (exports.hotp(options) === token) {
```

Comment:
Not off the top of my head, assuming that `parseInt(•, 10)` doesn’t have any major flaws.

> On Jan 24, 2016, at 11:58 AM, Mark Bao <notifications@github.com> wrote:
> 
> In index.js <https://github.com/speakeasyjs/speakeasy/pull/44#discussion_r50641088>:
> 
> > +
> > +exports.hotp.verifyDelta = function hotpVerifyDelta (options) {
> > +  var i;
> > +
> > +  // shadow options
> > +  options = Object.create(options);
> > +
> > +  // unpack options
> > +  var token = options.token;
> > +  var window = parseInt(options.window || 0, 10);
> > +  var counter = parseInt(options.counter || 0, 10);
> > +
> > +  // loop from C to C + W
> > +  for (i = counter; i <= counter + window; ++i) {
> > +    options.counter = i;
> > +    if (exports.hotp(options) === token) {
> Got it. Are there any issues you would think would arise from comparing converted ints instead of strings? I can't think of any right now, considering that front zero-padding shouldn't make a difference.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/speakeasyjs/speakeasy/pull/44/files#r50641088>.
>